main/src/library/List.flix

```diff
@@ -1096,21 +1096,44 @@ namespace List {
         }
 
     ///
-    /// Sort list `xs` with the comparing function `cmp`.
+    /// Sort list `xs` so that elements are ordered from low to high according to their `Order` instance.
+    ///
+    /// The sort is stable, i.e., equal elements appear in the same order as in the input `xs`.
+    ///
+    /// The sort implementation is a mergesort.
+    ///
+    pub def sort(xs: List[a]): List[a] with Order[a] = sortWith(Order.compare, xs)
+
+    /// Sort list `xs` so that elements are ordered from low to high according to the `Order` instance
+    /// for the values obtained by applying `f` to each element.
+    ///
+    /// The sort is stable, i.e., equal elements appear in the same order as in the input `xs`.
+    ///
+    /// The sort implementation is a mergesort.
+    ///
+    pub def sortBy(f: a -> b, xs: List[a]): List[a] with Order[b] = sortWith(Order.compare `on` f, xs)
+
+    ///
+    /// Sort list `xs` so that elements are ordered from low to high according to the comparison function `cmp`.
+    ///
+    /// The sort is stable, i.e., equal elements appear in the same order as in the input `xs`.
     ///
     /// The sort implementation is a mergesort.
+    ///
     /// Acknowledgement: derived from Thomas Nordin's `sortBy` in the Haskell base libraries.
     ///
-    @Time(length(xs) * Int32.log2(length(xs))) @Space(length(xs))
-    pub def sortWith(cmp: (a,a) -> Int32, xs: List[a]): List[a] =
+    /// TODO Review to verify it satisfies stability. The referenced `sortBy` is stable and
+    /// the tests suggest that this implementation is also.
+    ///
+    pub def sortWith(cmp: (a,a) -> Comparison, xs: List[a]): List[a] =
```

Comment:
Question: Perhaps we should simply convert the list to an array, sort it, and convert it back to a list. Apparently this is what Java does for most collections (!).

Reply:
Yes, that is also what I thought of (see also arguments in [#1945](https://github.com/flix/flix/issues/1945#issuecomment-843819838) and good point with Java. One would loose stability, but that is maybe not so often of interest anyway. If needed, we could add alternative stable sort functions. In that case, it should be reconsidered whether converting to an array and back is still more efficient for lists or whether for example mergesort on lists directly is better.

Reply:
OK. I should I merge now and then we can do that change later? (Need a new ticket?).

Reply:
I could also just add it here if we agree to change it; Maybe rename the current mergesort to `sortStable` and implement the actual `sort` using `toArray` and `toList`?

Reply:
https://github.com/flix/flix/issues/2003